source/core/slang-relative-container.h

```diff
@@ -0,0 +1,254 @@
+// slang-relative-container.h
+#ifndef SLANG_RELATIVE_CONTAINER_H_INCLUDED
+#define SLANG_RELATIVE_CONTAINER_H_INCLUDED
+
+#include "slang-basic.h"
+
+namespace Slang {
+
+struct RelativeBase
+{
+    uint8_t* m_data;
+
+    static RelativeBase g_null;
+};
+
+template <typename T>
+class Safe32Ptr
+{
+public:
+    typedef Safe32Ptr ThisType;
+
+    T& operator*() const { return *get(); }
+    T* operator->() const { return get(); }
+    operator T*() const { return get(); }
+
+    const Safe32Ptr& operator=(const ThisType& rhs) { m_offset = rhs.m_offset; m_base = rhs.m_base; return *this; }
+    SLANG_FORCE_INLINE T* get() const { return (T*)(m_base->m_data + m_offset); } 
+
+    void setNull()
+    {
+        m_offset = 0;
+        m_base = &RelativeBase::g_null; 
+    }
+
+    Safe32Ptr(const ThisType& rhs) : m_offset(rhs.m_offset), m_base(rhs.m_base) {}
+    Safe32Ptr() : m_base(&RelativeBase::g_null), m_offset(0) {}
+    Safe32Ptr(uint32_t offset, RelativeBase* base) : m_offset(offset), m_base(base) {}
+
+    RelativeBase* m_base;
+    uint32_t m_offset;
+};
+
+
+enum
+{
+    kRelative32PtrNull = int32_t(0x80000000)
+};
+
+template <typename T>
+class Relative32Ptr
```

Comment:
One thing I'll note with using types like this (which I've tried to do a few times before): there's definitely some stuff where if you don't have all the constructors and copy operations defined correctly you can create difficult-to-debug problems where the offset ends up being relative to some other C++ object because of implicit copies that get inserted.

I trust you to do a better and more complete job of this than me, but it is still something that makes me wary of this kind of type.

Reply:
Oh, and I just caught on now that the presence of `32` in the name and the lack of a `64` equivalent is because we are limiting the total size of the blobs, and not worrying about relative pointers across the whole address space. This also makes the encoding of null as a non-zero value work out fine. Got it.

Could still benefit from a comment on the `class` to lay all that out. :)

Reply:
On the first point - yes usage of these types have a bunch of fairly nasty restrictions/issues, particularly during construction. I will write this up on the types at some point, and where appropriate add checks/systems to ensure correct usage.